### PR TITLE
Storage refactoring

### PR DESCRIPTION
**Pull request by nupplaphil**
Splitting the `IWritableStorage`:

- `IWritableStorage` for CRUD operations at the storage backend
- `IStorageConfiguration` to configure the storage backend

@MrPetovan  It wasn't **that** easy because of the dynamic option fetching and saving, so I had to adapt some parts of `Core\StorageManager` and `Module\Admin\Storage` as well

This PR is a dependency for https://github.com/friendica/friendica-addons/pull/1162
FollowUp #10582

### CODE REVIEW

**Comment by MrPetovan** [comment]
> @MrPetovan It wasn't **that** easy because of the dynamic option fetching and saving, so I had to adapt some parts of `Core\StorageManager` and `Module\Admin\Storage` as well

And yet you still enjoyed it, didn't you?

File: doc/AddonStorageBackend.md [hunk 1]
```diff
@@ -31,7 +33,22 @@ interface IWritableStorage
 - `put(string $data, string $reference)` saves data in `$data` to position `$reference`, or a new position if `$reference` is empty.
 - `delete(string $reference)` delete data pointed by `$reference`
 
+### `IStorageConfiguration`
+
 Each storage backend can have options the admin can set in admin page.
+To make the options possible, you need to implement the `Friendica\Model\Storage\IStorageConfiguration` interface.
+
+All methods in the interface mus be implemented:
```
Comment (MrPetovan) [inline doc/AddonStorageBackend.md]: Isn't this a hard requirement for the class file to even parse after all?
```suggestion
All methods in the interface must be implemented:
```

File: doc/AddonStorageBackend.md [hunk 2]
```diff
@@ -278,29 +326,32 @@ The file is `addon/samplestorage/samplestorage.php`
...
-    }
+	$config          = new SampleStorageBackendConfig(DI::l10n(), DI::config());
+	$data['storage'] = new SampleStorageBackendConfig($config->getFileName());
 }
+
+function webdav_storage_config(App $a, array &$data)
+{
+	$data['storage_config'] = new WebDavConfig(DI::l10n(), DI::config());
+}
+
```
Comment (MrPetovan) [inline doc/AddonStorageBackend.md]: Projecting much? 😄

Let's stay in the `samplestorage` addon context.

File: doc/Addons.md [hunk 3]
```diff
@@ -802,6 +802,7 @@ Here is a complete list of all hook callbacks with file locations (as of 24-Sep-
 ### src/Core/StorageManager
 
     Hook::callAll('storage_instance', $data);
+	Hook::callAll('storage_config', $data);
```
Comment (MrPetovan) [inline doc/Addons.md]: Indentation 😉
```suggestion
    Hook::callAll('storage_config', $data);
```

File: src/Model/Storage/Filesystem.php [hunk 4]
```diff
@@ -40,30 +38,17 @@ class Filesystem implements IWritableStorage
...
+	 * @param string $filesystemPath
 	 */
-	public function __construct(IConfig $config, L10n $l10n)
+	public function __construct(string $filesystemPath = FilesystemConfig::DEFAULT_BASE_FOLDER)
 	{
-		$this->config = $config;
-		$this->l10n   = $l10n;
-
-		$path           = $this->config->get('storage', 'filesystem_path', self::DEFAULT_BASE_FOLDER);
+		$path           = $filesystemPath;
```
Comment (MrPetovan) [inline src/Model/Storage/Filesystem.php:53]: Shouldn't we check we have writing permission in the provided path?

File: doc/Addons.md (hunk 3 above)
Comment (MrPetovan) [inline doc/Addons.md]: Additionally, please describe the hook, its input and output keys in the `$data` array earlier in this file.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Review by MrPetovan — APPROVED** [review]

**Comment by MrPetovan** [comment]
And how the tests are failing with the newly created Exception!

**Comment by nupplaphil** [comment]
I love my tests :X :D

**Review by MrPetovan — APPROVED** [review]

**Pull request merged**